docs/spec/distributing.rst

```diff
@@ -10,72 +10,428 @@ Stub files
+
+    def foo(x: Literal[""]) -> int: ...
+
+Comments
+""""""""
+
+Standard Python comments are accepted everywhere Python syntax allows them.
+
+Two kinds of structured comments are accepted:
+
+* A ``# type: X`` comment at the end of a line that defines a variable,
+  declaring that the variable has type ``X``. However, :pep:`526`-style
+  variable annotations are preferred over type comments.
+* An error suppression comment at the end of any line. Common error suppression
+  formats are ``# type: ignore``, ``# type: ignore[error-class]``, and
+  ``# pyright: ignore[error-class]``. Type checkers may ignore error
+  suppressions that they don't support but should not error on them.
+
+Imports
+"""""""
+
+Stub files distinguish between imports that are re-exported and those
+that are only used internally. See :ref:`import-conventions`.
+
+Type aliases can be used to re-export an import under a different name::
+
+    from foo import bar as _bar
+    new_bar = _bar  # "bar" gets re-exported with the name "new_bar"
+
+Stubs support customizing star import semantics by defining a module-level
+variable called ``__all__``. In stubs, this must be a string list literal.
+Other types are not supported. Neither is the dynamic creation of this
+variable (for example by concatenation).
+
+When ``__all__`` is defined, exactly those names specified in ``__all__`` are
+imported::
+
+    __all__ = ['public_attr', '_private_looking_public_attr']
+
+    public_attr: int
+    _private_looking_public_attr: int
+    private_attr: int
+
+Type checkers support cyclic imports in stub files.
+
+Built-in Generics
+"""""""""""""""""
+
+:pep:`585` built-in generics are supported and should be used instead
+of the corresponding types from ``typing``::
+
+    from collections import defaultdict
+
+    def foo(t: type[MyClass]) -> list[int]: ...
+    x: defaultdict[int]
+
+Using imports from ``collections.abc`` instead of ``typing`` is
+generally possible and recommended::
 
-* Extension modules
+    from collections.abc import Iterable
 
-* Third-party modules whose authors have not yet added type hints
+    def foo(iter: Iterable[int]) -> None: ...
 
-* Standard library modules for which type hints have not yet been
-  written
+Unions
+""""""
+
+Declaring unions with the shorthand syntax or ``Union`` and ``Optional`` is
+supported by all type checkers::
+
+    def foo(x: int | str) -> int | None: ...  # recommended
+    def foo(x: Union[int, str]) -> Optional[int]: ...  # ok
+
+Module Level Attributes
+"""""""""""""""""""""""
+
+Module level variables and constants can be annotated using either
+type comments or variable annotation syntax::
+
+    x: int  # recommended
+    x: int = 0
+    x = 0  # type: int
+    x = ...  # type: int
+
+The ellipsis literal can stand in for any value::
+
+    x: int = ...  # type is int
+
+A variable annotated as ``Final`` and assigned a literal value has the
+corresponding ``Literal`` type::
+
+    x: Final = 0  # type is Literal[0]
```

Comment:
What types of value expressions should a type checker be expected to evaluate here? For example, should simple unary and binary arithmetic operators be supported? What about bitwise operators? What about the use of other `Final` variables? Mypy supports all of these, I think. Pyright recently filled out this full matrix.

```python
a: Final = -1
b: Final = 1 + 2
c: Final = 1 << 3
d: Final = 1 | 2 | 64
e: Final = (1 + c) ** 2
```

Reply:
What value is added by supporting things like arithmetic operators? My preference would be to allow only literal ints, strings, etc., on the principle that stubs shouldn't contain any unnecessary implementation detail.

Reply:
I think the value added is sometimes clarity for readers, e.g. `SECONDS_IN_DAY: Final = 60 * 60 * 24` vs `SECONDS_IN_DAY: Final = 86400`. You can add the clarity in a comment, but then it risks getting out of date with the code.

I don't think the value is large, but I suspect the cost is also not large, since I would expect a type checker to support this in non-stubs anyway. 

Though I note that the spec for Literal currently says nothing about support for arithmetic on integer literals, so perhaps we should avoid saying too much here, and defer to that spec, and it can be expanded in a separate PR with more clarity.

I would favor simply saying here that inference of Final literals in stubs should match inference of Final literals in non-stubs.

Reply:
I removed the description of how the inference works and linked to the appropriate section of the Literal spec.